### PR TITLE
fix: correct issue with funnels queries, switch to experimental analyzer

### PR DESCRIPTION
**Pull request by aspicer**
## Problem
**tl:dr; we should enable the [experimental analyzer](https://clickhouse.com/blog/clickhouse-release-24-03#analyzer-enabled-by-default) in Clickhouse.**

Why? Because we're getting bad data because the old analyzer allows bad queries, and who knows what it returns.
[Here is a slightly modified version of a query we're running](https://metabase.prod-eu.posthog.dev/question/281-analyzer). It's a funnel actors query. It runs, but it returns 5 rows. This isn't right, we were losing people, this is why I'm investigating it.

If you then enable the setting `allow_experimental_analyzer=1` at the end, it throws an error:
`Code: 215. DB::Exception: Column max_steps is not under aggregate function and not in GROUP BY keys.`
Hmmmm. So when I look at the code, I realize that the new analyzer is correct. We have a window function and we're trying to use it in a HAVING clause after a GROUP BY without a proper aggregation.

So I change
`HAVING ifNull(equals(steps, max_steps), isNull(steps) and isNull(max_steps)))`
to
`HAVING ifNull(equals(steps, max(max_steps)), isNull(steps) and isNull(max(max_steps))))`
and boom goes the dynamite - now it returns the proper amount, 20 rows. (edited)

## Changes

Enable experimental analyzer for funnels and fix query. Also add "hash" join as a fallback join type because this exception gets thrown otherwise.
```E DB::Exception: Only `hash` join supports multiple ORs for keys in JOIN ON section. Stack trace:```

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Passed tests and rigor of new analyzer.

### CODE REVIEW

**Comment by aspicer** [comment]
@webjunkie 

Would you mind taking a look at this failing test? I don't think I understand it. It's `test_funnel_step_breakdown_empty`
https://github.com/PostHog/posthog/actions/runs/9720856763/job/26832823920?pr=23348

Thanks!

**Comment by webjunkie** [comment]
> @webjunkie
> 
> Would you mind taking a look at this failing test? I don't think I understand it. It's `test_funnel_step_breakdown_empty` [PostHog/posthog/actions/runs/9720856763/job/26832823920?pr=23348](https://github.com/PostHog/posthog/actions/runs/9720856763/job/26832823920?pr=23348)
> 
> Thanks!

The idea was to have a query that returns some breakdown values as None... hence my idea to simulate this somehow. Maybe the clickhouse function works differently now with the adjusted analyzer? Any other ideas to get None as well as other legitimate breakdown values out?

**Comment by aspicer** [comment]
@webjunkie 

This change causes the query to return '' instead of None. Looking at [the issue,](https://github.com/PostHog/posthog/issues/22928) the empty string seems to actually be consistent with what we want, so this seems okay.

image

**Review by webjunkie — APPROVED** [review]

File: posthog/clickhouse/client/execute.py [hunk 1]
```diff
@@ -41,8 +41,11 @@
 
 @lru_cache(maxsize=1)
 def default_settings() -> dict:
+    # https://clickhouse.com/blog/clickhouse-fully-supports-joins-how-to-choose-the-right-algorithm-part5
+    # We default to three memory bound join operations, in decreasing order of speed
+    # The merge algorithms are not memory bound, and can be selectively used in places where it makes sense
     return {
-        "join_algorithm": "direct,parallel_hash",
+        "join_algorithm": "direct,parallel_hash,hash",
```
Comment (thecatontheflat) [inline posthog/clickhouse/client/execute.py:48]: Would it make sense to also allow it to spill on disk with the `partial_merge`?
